### PR TITLE
Add tokio

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

**Comment by piscisaureus** [comment]
@ry, @robbym PTAL

File: build_extra/rust/BUILD.gn [hunk 1]
```diff
@@ -143,5 +170,308 @@ rust_component("winapi") {
...
+    ":iovec",
+    ":kernel32",
+    ":lazycell",
+    ":libc",
+    ":log",
+    ":miow",
+    ":net2",
+    ":slab",
+  ]
+  extern_version = [
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Above each invocation of `extern_version` add a TODO. 
```
# TODO extern_version is for temporary workarounds only. Remove once deps use the same winapi version.
```

File: build_extra/rust/BUILD.gn [hunk 2]
```diff
@@ -143,5 +170,308 @@ rust_component("winapi") {
...
+    "ws2ipdef",
+    "ws2tcpip",
+    "wtypes",
+    "wtypesbase",
+  ]
+}
+
+rust_component("kernel32") {
+  source_root = "$registry_github/kernel32-sys-0.2.2/src/lib.rs"
+  extern_version = [
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Above each invocation of `extern_version` add a TODO. 
```
# TODO extern_version is for temporary workarounds only. Remove once deps use the same winapi version.
```

File: build_extra/rust/BUILD.gn [hunk 3]
```diff
@@ -143,5 +170,308 @@ rust_component("winapi") {
...
+}
+
+rust_component("miow") {
+  source_root = "$registry_github/miow-0.2.1/src/lib.rs"
+  extern = [
+    ":kernel32",
+    ":net2",
+    ":ws2_32",
+  ]
+  extern_version = [
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Above each invocation of `extern_version` add a TODO. 
```
# TODO extern_version is for temporary workarounds only. Remove once deps use the same winapi version.
```

File: build_extra/rust/BUILD.gn [hunk 4]
```diff
@@ -143,5 +170,308 @@ rust_component("winapi") {
...
+      crate_name = "winapi"
+      crate_version = "0.2"
+    },
+  ]
+}
+
+rust_component("iovec") {
+  source_root = "$registry_github/iovec-0.1.2/src/lib.rs"
+  extern = [ ":libc" ]
+  extern_version = [
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Above each invocation of `extern_version` add a TODO. 
```
# TODO extern_version is for temporary workarounds only. Remove once deps use the same winapi version.
```

File: build_extra/rust/BUILD.gn [hunk 5]
```diff
@@ -143,5 +170,308 @@ rust_component("winapi") {
...
+  source_root = "$registry_github/net2-0.2.33/src/lib.rs"
+  features = [
+    "default",
+    "duration",
+  ]
+  extern = [
+    ":cfg_if",
+    ":libc",
+  ]
+  extern_version = [
```
Comment (ry) [inline build_extra/rust/BUILD.gn]: Above each invocation of `extern_version` add a TODO. 
```
# TODO extern_version is for temporary workarounds only. Remove once deps use the same winapi version.
```

File: BUILD.gn [hunk 6]
```diff
@@ -50,19 +51,60 @@ main_extern = [
...
+  "$rust_build:tokio_tcp",
+  "$rust_build:tokio_threadpool",
+  "$rust_build:tokio_current_thread",
+  "$rust_build:tokio_timer",
+  "$rust_build:tokio_udp",
+  "$rust_build:unicode_bidi",
+  "$rust_build:unicode_normalization",
+  "$rust_build:ws2_32",
+]
+main_extern_version = [
```
Comment (ry) [inline BUILD.gn]: Add TODO about extern_version being a hack here too.

**Comment by ry** [comment]
Could you also add an issue for merging the multiple winapi versions and link to it in the TODOs requested above?

File: build_extra/rust/rust.gni [hunk 7]
```diff
@@ -137,48 +152,61 @@ template("run_rustc") {
...
-                             "extern",
-                             "features",
-                             "is_test",
-                             "source_root",
-                             "testonly",
-                           ])
-    if (!defined(invoker.crate_type)) {
-      crate_type = "rlib"
-    }
+    forward_variables_from(invoker, "*")
```
[the remark this answers is not in the document]
Comment (piscisaureus) [reply build_extra/rust/rust.gni]: Fixed. I didn't realize this. Maybe we should add an OS X build to Travis?

[the remark this answers is not in the document]
Comment (ry) [reply build_extra/rust/rust.gni]: Eventually.

**Pull request merged**